### PR TITLE
`Assessment`: Allow instructors to export course scores to an Excel file

### PR DESCRIPTION
**Pull request by maximiliansoelch**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Currently Artemis only allows to export course scores to a CSV file. When a user specifies another decimal separator than their system's setting and then imports the CSV file into Excel, values are interpreted differently (e.g. Custom Decimal Separator: '.', User system set to german: ',' -> Generated csv contains numbers in the format 1.4 -> Excel interprets this as a date (1. April)).
In addition, allowing to export directly to an Excel file improves the usability for technical inexperienced users.

### Description

The introduced changes now allow an instructor to download course scores directly as an Excel file by utilizing the library [SheetJS](https://github.com/SheetJS/sheetjs).

### Steps for Testing

Prerequisites:

- 1 Course
- 1 Student
- 1 Exercise, on which the student participated

1. Log in to Artemis
2. Navigate to Course Administration, select the course, then Scores
3. Click the `Export` button
4. A modal opens, where you can select the file type used for the export (Excel<default> or CSV).
5. Select Excel (which should be selected initially) and click `Export as Excel`
6. Open the downloaded Excel file and make sure the rows and their values are equivalent to the ones displayed in Artemis
7. In addition, also verify that the export to a CSV file still works as expected.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Branch | Line |
|------------|-------:|-----:|
| course-scores-csv-row-builder.ts | 100% | 100% |
| course-scores-excel-row-builder.ts | 100% | 100% |
| course-scores-row-builder.ts | 96% | 100% |
| course-scores.component.ts | 73.52% | 93.17% |
| export-modal.component.ts | 100% | 100% |

### Screenshots

Excel_Export

### CODE REVIEW

File: src/main/webapp/app/course/course-scores/course-scores.component.ts [hunk 1]
```diff
@@ -508,36 +512,59 @@ export class CourseScoresComponent implements OnInit, OnDestroy {
...
-        rows.push(this.generateCsvRowAverageValues(customOptions.decimalSeparator));
-        rows.push(this.generateCsvRowParticipation(customOptions.decimalSeparator));
-        rows.push(this.generateCsvRowSuccessfulParticipation(customOptions.decimalSeparator));
+        rows.push(this.generateCsvRowMaxValues(customCsvOptions));
+        rows.push(this.generateCsvRowAverageValues(customCsvOptions));
+        rows.push(this.generateCsvRowParticipation(customCsvOptions));
+        rows.push(this.generateCsvRowSuccessfulParticipation(customCsvOptions));
 
-        this.exportAsCsv(quotedKeys, rows, customOptions);
+        if (!!customCsvOptions) {
```
Comment (bassner) [inline src/main/webapp/app/course/course-scores/course-scores.component.ts:533]:
```suggestion
        if (customCsvOptions) {
```

File: src/main/webapp/app/course/course-scores/course-scores.component.ts [hunk 2]
```diff
@@ -508,36 +512,59 @@ export class CourseScoresComponent implements OnInit, OnDestroy {
...
+     */
+    exportAsExcel(keys: string[], rows: CourseScoresExportRow[]) {
+        const workbook = XLSX.utils.book_new();
+        const ws = XLSX.utils.json_to_sheet(rows, { header: keys });
+        const worksheetName = `${this.course.title} Scores`;
+        XLSX.utils.book_append_sheet(workbook, ws, worksheetName);
+
+        const workbookProps = {
+            Title: `${this.course.title} Scores`,
+            Author: `Artemis ${VERSION ? VERSION : ''}`,
```
Comment (bassner) [inline src/main/webapp/app/course/course-scores/course-scores.component.ts:555]:
```suggestion
            Author: `Artemis ${VERSION ?? ''}`,
```

**Review by bassner — CHANGES_REQUESTED** [review]
I've tested this on TS1 and was unable to export an Excel file due to this error in the console:
grafik
(Tried on https://artemis-test1.artemis.in.tum.de/course-management/7/scores)

Additionally, please see my few code remarks below.

**Comment by maximiliansoelch** [comment]
@bassner Thanks for testing. I will look into the error ...

**Comment by maximiliansoelch** [comment]
@bassner The error you observed should be fixed now. I also implemented your code suggestions.

File: src/main/webapp/app/shared/export/export-modal.component.ts [hunk 3]
```diff
@@ -30,14 +30,15 @@ export interface CsvExportOptions {
...
+    templateUrl: './export-modal.component.html',
+    styleUrls: ['./export-modal.component.scss'],
 })
-export class CsvExportModalComponent implements OnInit {
+export class ExportModalComponent implements OnInit {
     readonly CsvFieldSeparator = CsvFieldSeparator;
     readonly CsvQuoteStrings = CsvQuoteStrings;
     readonly CsvDecimalSeparator = CsvDecimalSeparator;
 
+    active = 1;
```
Comment (ge65cer) [inline src/main/webapp/app/shared/export/export-modal.component.ts:41]: I'd rename this `activeTab` or similar to give better context

File: src/main/webapp/app/course/course-scores/course-scores-excel-row-builder.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import { CourseScoresRowBuilder } from 'app/course/course-scores/course-scores-row-builder';
+import { round } from 'app/shared/util/utils';
+
+/**
+ * This interface represents a cell object, which conforms to the Common Spreadsheet Format (CSF).
+ * https://github.com/sheetjs/sheetjs#common-spreadsheet-format
+ */
+export interface CommonSpreadsheetCellObject {
```
Comment (ge65cer) [inline src/main/webapp/app/course/course-scores/course-scores-excel-row-builder.ts:8]: Maybe you could add the description for the attributes as comments directly here?

**Review by ge65cer — COMMENTED** [review]
Great work on the functionality, tested on TS2 and worked correctly.
I've left some small remarks.

**Comment by maximiliansoelch** [comment]
@bassner Yes, for now that is intended as Artemis stores grades as strings.

**Review by ge65cer — APPROVED** [review]
Thanks for the small changes, looks good to me now!

**Review by joaquinmurcia — CHANGES_REQUESTED** [review]
Tested with test_user_2 on TS2. The columns in the exported (both Excel and CSV) files don't match the columns shown in Artemis. Is this supposed to happen?
image
image

**Comment by maximiliansoelch** [comment]
@joaquinmurcia Thanks for testing.
Yeah, it is intended that the order of the columns is not the same (e.g. in Artemis the _avg_ and _max_ column are located at the top, in the export files they are located at the bottom).

What is important is that the exported points and scores of the students match the data in Artemis as well as the values for _avg_ and _max_ (what they do in your posted case).

**Review by max-bergmann — DISMISSED** [review]
Tested on TS3. Works as expected.

**Review by jschmautzer — APPROVED** [review]
Tested on ts1, followed the described testing steps, works as expected :)

**Review by ge65cer — APPROVED** [review]
Re-approve

**Review by willergeorg — APPROVED** [review]
Tested on TS3. The scores in the Excel file match the values displayed by Artemis.

**Review by jschmautzer — APPROVED** [review]
re-approve, tested on ts1

**Pull request merged**